[Proposal] Additional rules for Danger and SwiftLint

NSMyself opened the pull request:
* [x] I will send a meeting invitation, using this [template](https://raw.githubusercontent.com/Babylonpartners/ios-playbook/master/Cookbook/Proposals/Template_Proposal_Meeting_Invitation.md), scheduled for 2 weeks after this proposal is made, so an agreement can be reached.
* [x] **By creating this proposal, I understand that it might not be accepted**. I also agree that, if it's accepted,
depending on its complexity, I might be requested to give a workshop to the rest of the team. 🚀

Commit:
Initial draft

AliSoftware reviewed (commented):
Please give more details about the rules and the specs about them.
I think for most of the rules many will agree they could be useful so there won't be much debate probably, but there's a big need for discussion about the details specs.

For example if you pitch a rule about detecting that Current should only be used in builders, if you dig a little deeper (which is the point of writing a proposal, for the author to think about all the implications and the solution in more details) you realize that it's a little too light to describe the exact rule to implement and that we need to specify it more in depth. Eg we should make sure it's not triggered in AppDelegate as Current is initialized here, probably AppConfiguration too… should we whitelist or blacklist…

Having a quick draft, even if not perfect, of the danger or SwiftLint rule and how it will be shaped helps a lot forcing you to think more in depth about all those questions that are the real things we'll want to discuss when reviewing the proposal as a team.

Commit:
Minor fixes

Commit:
Improve lifecycle monitoring rule

Commit:
Add

Commit:
Add when rule

Commit:
Add  rule

Commit:
WIP Current mutations

Commit:
Fix Current mutation rule

Commit:
Cleanup

MichalTKwiecien commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,131 @@
+### Deprecated calls of DesignLibrary components (aka pre-NVL components)
+
+Given that NVL is scheduled for release within two weeks, using NVL components is pivotal to Babylon's overall look and feel.
+With the exception of the empty space component (even though it has its own NVL equivalent), every other pre-NVL component should trigger a warning.
```
I think we were using this because there was no NVL equivalent - if we have it now let's just switch to it and do not make any exceptions.

MichalTKwiecien commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,131 @@
+### Deprecated calls of DesignLibrary components (aka pre-NVL components)
+
+Given that NVL is scheduled for release within two weeks, using NVL components is pivotal to Babylon's overall look and feel.
+With the exception of the empty space component (even though it has its own NVL equivalent), every other pre-NVL component should trigger a warning.
```
Should only be checking `Renderer`s for those or ignoring `BabylonUI` to no trigger warnings for proper usage.

```suggestion
With the exception of the empty space component (even though it has its own NVL equivalent), every other pre-NVL component used inside Renderer should trigger a warning.
```

NSMyself replied:
Well said, thanks 👍

NSMyself committed:
Update Cookbook/Proposals/AdditionalRulesForDanger.md

Co-Authored-By: Michał Kwiecień <michal.t.kwiecien@gmail.com>

NSMyself committed:
Update AdditionalRulesForDanger.md

Improve `motivation` section

NSMyself committed:
Update AdditionalRulesForDanger.md

dmcrodrigues commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+
+In order to detect this, we could simply monitor occurences of the following regex:
+
+`\.signal\(for: \#selector\(UIViewController\.`
```
This only addresses part the problem, we have a reasonable number of custom view controllers doing a similar logic too.

dmcrodrigues commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+The algorithm is then be executed once more but for the `tearDown()` method instead.
+If the number of changes inside each function matches, then no issue would be found. If not, then Current appeared to have been mutated but not restored; we'd issue a warning for the lines in which the algorithm detected a change and alert the developers in GitHub.
+
+Nevertheless, this rule is rather naïve because we have no actual way of detecting if Current was reset or not; that would require compiling and runniing the actual Swift code. Therefore, this approach is not a silver bullet or anything remotely similar but it would still be useful for one use case in particular: should the developer forget to restore Current, this rule will alert us of this situ ...
```
```suggestion
Nevertheless, this rule is rather naïve because we have no actual way of detecting if Current was reset or not; that would require compiling and running the actual Swift code. Therefore, this approach is not a silver bullet or anything remotely similar but it would still be useful for one use case in particular: should the developer forget to restore Current, this rule will alert us of this situation.
```

dmcrodrigues commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+
+We would need to whitelist `Localizable.*\.swift` in order to make this work.
+
+### Feedbacks must have the `when` prefix
```
I'm not sure about this one, we have Feedbacks which start their effects when the system is initialised and not depending on a certain state so I think it would be misleading enforcing this prefix to them.

ilyapuchka replied:
@dmcrodrigues then it can be called `whenInitialized` or `whenStarted` or something like that? I'm +1 for enforcing common naming patterns in general. Similarly it would be nice to enforce common pattern in Actions/Events, i.e. that they start or should not start with `did` and they don't match state names. But it might be trickier as we are dealing with natural language processing here and a simple regex can be too limiting.

dmcrodrigues reviewed (approved):
Looks good, I think it would be beneficial to have this set of rules to kind of enforce but also promote our own conventions.

NSMyself committed:
Update Cookbook/Proposals/AdditionalRulesForDanger.md

Co-Authored-By: David Rodrigues <david.ndh@gmail.com>

chitrakotwani commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+
+
+### Mutating Current in `setUp()` but not restoring it in `tearDown()`
+
```
I was under the impression that `Current` should not be done at all even if it is restored in `tearDown()` 

While restoring it in `tearDown()` mitigates the problem, it does not eliminate it.

The sequence of events can be : 
`Test1.setUp(){Set UK locale on Current}`
`Test2.setUp(){Set USA locale on Current}` 
`Test1.test_xyz  {Expect time to be rendered according to UK locale} `  ❌ 

I think this rule can be changed to not allow mutating Current at all.

chitrakotwani commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+The sheer number of warnings immediately excludes SwiftLint; Danger would be our only option.
+
+Nevertheless, this would require that everyone relied exclusively on NVL components and I am not sure this is the case right now.
+
```
Enforcing `Danger` to apply this rule will be too late. Significant effort and time will have been used till then to create a screen with deprecated components. (With my limited understanding of what NVL encompasses)

NSMyself replied:
Good point. I will adjust the proposal, this seems like a good trade-off.

chitrakotwani commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+
+### Custom ViewModels in the test suite
+
+Custom ViewModels should not be defined as part of our tests; we should rely on `StubViewModel`s (from `BabylonSnapshotTestUtilities`) instead (although, in theory, setting up custom ViewModels could be useful in particular circumstances).
```
I don't have much context here. Why and what are these? In which circumstances can they be useful?

ilyapuchka replied:
We used to define stubs for view models in the snapshot tests as they are required to create the screens. Now we have a generic `StubViewModel` for that and we don't need custom implementations, but they are still being used somewhere. I'd prefer this to be solved with deprecation annotation though. We can for example write a variant of `verifyScreenForAllSizesAndVisualLanguages` that would only accept `StubViewModel` for `viewModel` parameter and deprecate the variant that is generic over type of ViewModel. @NSMyself  what do you think?

NSMyself replied:
@chitrakotwani apologies for not replying, I couldn't do it at the time and completely forgot about it. 

Basically, by using custom ViewModels instead of stubbing the ones we already have technically we aren't testing the actual ViewModels, but the ones we've just created for the purpose of this test. Depending on the actual implementation, that may not be the exact same thing.

Regarding the circumstances in which this approach could be useful, well:
a) we may need to test something that because of it's nature and/or complexity is not easily achievable and, as such, using a custom ViewModel could theoretically help (although that fact could indicate that the VM needs to be refactored, but that's another story).

b) a concrete example: when testing a Renderer and we need our VM to hold some value that we cannot define via the initializers and/or easily get the state machine to the point where we need it to be.

NSMyself replied:
@ilyapuchka i think that's a great idea. I need to revisit this proposal and update it because in some cases marking the method via a deprecation annotation is indeed the right approach imho. Will do so in the following days.

chitrakotwani reviewed (commented):
Much needed undertaking! 👏

AliSoftware commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+In order to detect this problem we have to use Danger since SwiftLint only accepts regular expressions.
+We would have to whitelist every test file (`.*Tests.*`) and use Danger to manually parse the test line-by-line until the `setUp()` function was found. 
+
+If found, we need to make a counter that increases by one each time a curly bracket and does the reverse when a closing curly bracket appears. Once the counter reaches zero again, then we've successfully delimited the contents of the `setUp()` function and we'd use the following regex to detect changes in Current: `Current\..*\ = `.
```
We need to also detect `Current.changeTheme` or other `Current.setXXX`, not just `Current = `

NSMyself replied:
Actually, @daniloaliberti pointed out that this issue is [already covered](https://github.com/Babylonpartners/babylon-ios/pull/7806/files#diff-74895f4da31ec40d83d342bf612540b4R17) 👍

I will remove this section entirely

NSMyself replied:
> Can I suggest to add the rule that would validate the PR metadata itself, its title at least 🙏 Currently we have a bot that will fail if there is no properly formatted ticket number in the title, but it does not explain the failure. We can extend this further to validating that if there are no test changes in the PR then there should be a note in the PR description about that

I think that's a good idea @ilyapuchka but we should perhaps make another proposal because this one already has a lot of stuff for just a 15m meeting

AliSoftware commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+
+As such, I'd like to suggest another rule for Danger that would:
+    1. detect new static String variables in every Localizable.swift file.
+    2. if we detected a new static variable we would then scan the localization string file (`Babylon/Supporting Files/Localization/en.lproj/Localizable.strings`) and look for that key
```
1. I think we need to do that in both directions. A key in `Localization*.swift` but not present in `Localizable.strings` is the biggest problem, but a key in the `.strings` file without a constant in the `.swift` file also means that we pulled a strings from Lokalise that we in fact don't use in this PR and is probably part of a different, unrelated feature, so we should warn about this too
2. What about other `.strings` files, especially `TargetSpecificLocalizable.strings`?

Also, I think we need to address the problem about handling the L10n across our modules and in tests. e.g. if that's what we've decided today (I'm not even sure what the actual concensus is today tbh), ensure that any reference to a `Localizable.strings` in a test target of a project file is just a reference to the `Localizable.strings` file of Babylon, so containing the same keys since it's the same file, and is not a separate `.strings` file where keys have been copy/pasted instead.

I think all that might even require a TechnicalDocument on its own jsut to describe how to do L10n on Babylon (using the dedicated lane to pull translations, clean up translations unrelated to the PR, in which files to add the translations, in which file to add the constants, depending on the feature they're related to (`LocalizationUI.swift` vs `LocalizationOctopus.swift`, etc), how to reference localized strings from tests by switching the bundle...

And once all those rules are clearly written black on white, the Danger rule will become much clearer too. e.g. I can already see a new rule we should have, similar to the one about `Current`, to ensure the bundle is reset in `tearDown` if it was changed in `setUp`...

AliSoftware replied:
I'm all for using a code generation tool for that.

But we'll need a lot of cleanup of our existing codebase first I think, since as of now, the name of the constants we manually create and the corresponding Localization.strings keys they point to don't match names. So when we'll replace that with codegen, the name of our constants will completely change to match the keys used in the `.strings` and we'll have a lot of updating to do on all the code using them. Which means it's not a trivial task given our legacy and number of existing keys (we might consider a progressive transition though)

Since idk when we'll get there (transition to codegen), that might be an intermediate rule worth having in the meantime

ilyapuchka replied:
we'll have to do that anyway. if we need an intermediate solution that will help with it so that we can do that gradually rather than all in one go I think it's good. But again even if we start to code generate these strings today I don't think anything enforces us to use generated code all in, we can keep existing structures and just mark it as deprecated. Lots of options here.

AliSoftware commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+
+* Implement a subset of those rules (and determine the most adequate tool to detect infractions for each rule).
+
+* Don't implement these rules at all
```
* Fix some of the existing warnings (maybe some of them are in great number but relatively simple to fix with search/replace for example?) first, so that we could then re-consider make new warnings like SwiftLint's be more visible and thus more useful.

* Use SwiftLint for some of the rules mentioned above, even if we already have violations for them in our current codebase, but keep them as warnings. Would increase the number of warnings again, but would also allow us to see warnings in any new code we write immediately, instead of only at PR time. Would also allow us to see warnings on existing code when we work on an existing file and hopefully take that occasion to refactor and fix the warning while working on our feature (e.g. if I work on an existing, legacy screen, if I only get the warning via Danger I might not be aware that it's legacy and I should take the occasion to refactor it a bit while I'm at it... until I've done my ticket and open the PR)

* Use SwiftLint and make the violations be error, but use `// swiftlint:disable` on existing code already violating the rules, so that only the new code gets the warnings. (I'm not fan of this solution, since it would risk to legitimize the use of `swiftlint:disable` and end up defeating the purpose of SwiftLint in the long run, but had to mention it for the sake of argument)

NSMyself replied:
Fair points, will revisit this section.

ilyapuchka reviewed (commented):
I'm generally +1 for the direction of ensuring that our codebase is uniform, some implementation details need to be discussed but otherwise good ideas 👍

ilyapuchka commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,133 @@
+### Monitoring ViewController lifecycle directly instead of relying on `ScreenLifecycleEvent`
+
+With the introduction of `ScreenLifecycleEvent`, developers should no longer manually monitor the viewController's lifecycle events.
+Instead, this kind of monitoring can now easily be done via the `send(_: ScreenLifecycleEvent)` method.
```
tbh I don't know what it is 🤷‍♂

ilyapuchka commented:
Can I suggest to add the rule that would validate the PR metadata itself, its title at least 🙏 Currently we have a bot that will fail if there is no properly formatted ticket number in the title, but it does not explain the failure. We can extend this further to validating that if there are no test changes in the PR then there should be a note in the PR description about that

Commit:
Improve proposal following PR feedback

Commit:
Fix merge conflicts

Commit:
WIP

Commit:
Wrap it up

Commit:
Add Danilo and Olivier as authors

NSMyself commented:
I've just up updated the proposal after considering feedback from @AliSoftware, @MichalTKwiecien, @dmcrodrigues, @chitrakotwani and @ilyapuchka.

Additionally, I also added another section after chatting with @daniloaliberti and @AliSoftware regarding monitoring the usage of Current outside builders. That section was made by the three of us, which was why I added them as authors as well.

Commit:
Update last paragraph

Commit:
Fix typo

daniloaliberti commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,208 @@
+
+### Detect Current usage outside Builders et al
+
+`Current` is a very powerful concept whose usage is only allowed in a very restrict number of places. As such, we should enable automatic monitoring in order to prevent it from being abused throughout our codebase. Nevertheless, Current is still being used in a lot of places (598 results across 144 files) and we should go through every case individually before commiting to a decision:
```
```suggestion
`Current` is a very powerful concept whose usage is only allowed in a very restrict number of places. 
Check [Current Guide](./Technical-Documents/Current-guide.md) . As such, we should enable automatic monitoring in order to prevent it from being abused throughout our codebase. Nevertheless, Current is still being used in a lot of places (598 results across 144 files) and we should go through every case individually before commiting to a decision:
```

daniloaliberti commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,208 @@
+
+Every object (susceptible to global mutation) that is mutated in `setUp()` must be restored in `tearDown()`.
+
+Mutations in `Current` in particular are the prime suspects of a number of flaky tests that have been recently affecting our test suite. [While some mitigations have been implemented to minimize this problem](https://github.com/Babylonpartners/babylon-ios/pull/7806/files#diff-74895f4da31ec40d83d342bf612540b4R17), there are still plenty of cases through which Current could be modified and our team ...
```
```suggestion
Mutations in `Current` in particular are the prime suspects of a number of flaky tests that have been recently affecting our test suite. [While some mitigations have been implemented to minimize this problem](https://github.com/Babylonpartners/babylon-ios/pull/7806/files#diff-74895f4da31ec40d83d342bf612540b4R17), there are still plenty of cases through which Current could be modified and our team would be none the wiser, like extending the linked solution to base Unit and UI Test cases.
```

NSMyself committed:
Update Cookbook/Proposals/AdditionalRulesForDanger.md

Co-Authored-By: Danilo Aliberti <danilo.aliberti@babylonhealth.com>

NSMyself committed:
Update Cookbook/Proposals/AdditionalRulesForDanger.md

Co-Authored-By: Danilo Aliberti <danilo.aliberti@babylonhealth.com>

NSMyself committed:
Update AdditionalRulesForDanger.md

NSMyself commented:
After considering the fact that our proposal meeting only lasts 15m I decided to move the new section (Current outside builders) to a new proposal

Commit:
Update AdditionalRulesForDanger.md

Remove `Current outside of builders` section

Commit:
Refactor setUp()tearDown()

Commit:
Minor edit

Commit:
Fix title

Commit:
Update AdditionalRulesForDanger.md

daniloaliberti commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,151 @@
+
+Furthermore, the reverse situation would also be problematic (albeit at a different scale): a key in the `.strings` file without an equivalent constant in its Swift counterpart would likely mean that the developer had added unused strings to the project. This would also be detectable using Danger.
+
+We would need to whitelist `Localizable.*\.swift` as well as the actual `Localizable.string` files in order to implement both rules.
```
let's not forget the `TargetSpecificLocalizable.strings`

daniloaliberti replied:
also let's not forget what @dmcrodrigues mentioned: SwiftGen to do not rely on developers

NSMyself replied:
@daniloaliberti @dmcrodrigues I added some tweaks regarding the localization rule, please have a look when you've time

NSMyself commented:
Outcome of this meeting:

`Detecting custom ViewModels in snapshot tests`: ✅
`ScreenLifecycleEvent instead of observing VC lifecycle directly`: ✅
`Deprecated calls of DesignLibrary components`: ✅
`Mutating Current in `setUp()` but not restoring it in `tearDown()`: ❌
`Feedbacks must have the `when` prefix`: ✅

Commit:
Update AdditionalRulesForDanger.md

daniloaliberti commented on Cookbook/Proposals/AdditionalRulesForDanger.md:
```diff
@@ -0,0 +1,155 @@
+- we have no reliable way of mapping the association between each target and their respective strings
+- this process should be automated by using SwiftGen instead
+
+We would need to whitelist `Localizable.*\.swift` as well as the `Localizable.string` and its corresponding target-specific files in order to implement both rules.
```
```suggestion
We would need to whitelist `Localizable.*\.swift` as well as the `Localizable.string` and its corresponding target-specific files (TargetSpecificLocalizable.strings) in order to implement both rules.
```

Commit:
Update Cookbook/Proposals/AdditionalRulesForDanger.md

Co-Authored-By: Danilo Aliberti <danilo.aliberti@babylonhealth.com>

Commit:
Strikethrough the Mutating Current rule

The `Mutating Current in setUp() but not restoring it in tearDown()` rule was strikethroughed because it was rejected during the meeting

daniloaliberti reviewed (approved):
great one man!
